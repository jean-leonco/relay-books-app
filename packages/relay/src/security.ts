import AsyncStorage from '@react-native-community/async-storage';

import { AUTH_TOKEN } from './config';

export async function getToken() {
  return AsyncStorage.getItem(AUTH_TOKEN);
}

export async function clearToken() {
  return AsyncStorage.removeItem(AUTH_TOKEN);
}
