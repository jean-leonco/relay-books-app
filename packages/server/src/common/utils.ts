import { Model, Types } from 'mongoose';
import { fromGlobalId } from 'graphql-relay';

// https://stackoverflow.com/a/2593661/710693
export const escapeRegex = (str: string) => `${str}`.replace(/[.?*+^$[\]\\(){}|-]/g, '\\$&');

export const PLATFORM = {
  APP: 'APP',
  UNKNOWN: 'UNKNOWN',
};

export const getPlatform = (platform?: string | null): string => {
  if (typeof platform !== 'string') {
    return PLATFORM.UNKNOWN;
  }

  //if (platform.includes('ios') || platform.includes('android')) {
  //  return PLATFORM.APP;
  //}

  if (platform === PLATFORM.APP) {
    return PLATFORM.APP;
  }

  return PLATFORM.UNKNOWN;
};

export function kilometersToEarthRadians(kilometers: number) {
  return kilometers / 6378.1;
}

// returns an ObjectId given an param of unknown type
export const getObjectId = (target: string | Model<any> | Types.ObjectId | undefined | null): Types.ObjectId | null => {
  if (!target) {
    return null;
  }

  if (target instanceof Types.ObjectId) {
    return new Types.ObjectId(target.toString());
  }

  if (typeof target === 'object') {
    return target && target._id ? new Types.ObjectId(target._id) : null;
  }

  if (typeof target === 'string') {
    const result = fromGlobalId(target);

    if (result.type && result.id && Types.ObjectId.isValid(result.id)) {
      return new Types.ObjectId(result.id);
    }

    if (Types.ObjectId.isValid(target)) {
      return new Types.ObjectId(target);
    }

    return null;
  }

  return null;
};

type EnumObject = {
  [key: string]: string;
};

type EnumObjectResult = {
  [key: string]: {
    value: string;
  };
};

export const enumBuilderValues = (constants: EnumObject): EnumObjectResult =>
  Object.keys(constants).reduce(
    (prev, curr) => ({
      ...prev,
      [curr]: {
        value: constants[curr],
      },
    }),
    {},
  );
