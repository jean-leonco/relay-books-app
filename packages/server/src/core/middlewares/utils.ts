import { GraphQLResolveInfo } from 'graphql';
import { ObjectSchema, Shape } from 'yup';

import { GraphQLContext } from '../../types';

interface GraphQLMutationConfig {
  validationSchema?(context: GraphQLContext): ObjectSchema<Shape<any, any>>;
  authenticatedOnly?: boolean;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): GraphQLMutationConfig | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
