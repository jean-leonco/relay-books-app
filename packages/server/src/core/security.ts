import { GraphQLContext, LoggedGraphQLContext } from '../types';

import { User } from '../loader';

export const isLoggedIn = (context: GraphQLContext | LoggedGraphQLContext) => {
  const { user } = context;

  if (user instanceof User) {
    return true;
  }

  return false;
};
