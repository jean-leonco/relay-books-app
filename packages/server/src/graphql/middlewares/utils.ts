import { GraphQLField, GraphQLResolveInfo } from 'graphql';

import { Extensions } from '../../types';

interface Field extends GraphQLField<any, any> {
  extensions: Extensions | null | undefined;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): Field | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
