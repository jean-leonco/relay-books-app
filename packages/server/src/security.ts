import User from './modules/user/UserLoader';

import { GraphQLContext, IStatusSchema } from './types';

export const isLoggedIn = (context: GraphQLContext) => {
  const { user } = context;

  if (user instanceof User) {
    return true;
  }

  return false;
};

export const isLoggedViewerCanSee = (context: GraphQLContext, data: IStatusSchema) => {
  if (!context.user || !data.isActive) {
    return false;
  }

  return true;
};

export enum PLATFORM {
  APP = 'APP',
  UNKNOWN = 'UNKNOWN',
}

export const getPlatform = (platform?: string | null) => {
  if (typeof platform !== 'string') {
    return PLATFORM.UNKNOWN;
  }

  //if (platform.includes('ios') || platform.includes('android')) {
  //  return PLATFORM.APP;
  //}

  if (platform === PLATFORM.APP) {
    return PLATFORM.APP;
  }

  return PLATFORM.UNKNOWN;
};
