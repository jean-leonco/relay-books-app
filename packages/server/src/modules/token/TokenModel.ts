import mongoose, { Document, Model } from 'mongoose';

import { IStatusSchema, ObjectId } from '../../types';

import { PLATFORM } from '../../security';
import isActiveMongooseField from '../../mongoose/isActiveMongooseField';

export enum TOKEN_SCOPES {
  RESET_PASSWORD = 'BOOKSAPP:RESET_PASSWORD', // used when user requested a password reset
  SESSION = 'BOOKSAPP:SESSION', // token for a valid user session
}

const Schema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      description: 'The user owner of this token.',
      index: true,
      default: null,
    },
    ip: {
      type: String,
      description: 'Agent IP that came from request headers.',
      required: true,
      index: true,
    },
    platform: {
      type: String,
      required: true,
      default: PLATFORM.APP,
      description: 'Channel that this token belongs to. ex: APP.',
      enum: Object.values(PLATFORM),
    },
    scope: {
      type: String,
      description: 'Scope of this token. ex: SESSION.',
      required: true,
      index: true,
      enum: Object.values(TOKEN_SCOPES),
    },
    expiresIn: {
      type: Date,
      description: 'When this token expires.',
      default: null,
    },
    isBlocked: {
      type: Boolean,
      description: 'If the token is blocked.',
      default: false,
      index: true,
    },
    ...isActiveMongooseField,
  },
  {
    timestamps: true,
    collection: 'Token',
  },
);

export interface IToken extends Document, IStatusSchema {
  userId?: ObjectId;
  ip: string;
  platform: PLATFORM;
  scope: TOKEN_SCOPES;
  expiresIn: Date | null;
  isBlocked: boolean;
}

const TokenModel: Model<IToken> = mongoose.model('Token', Schema);

export default TokenModel;
