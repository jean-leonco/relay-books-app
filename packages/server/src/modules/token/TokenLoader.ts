import DataLoader from 'dataloader';
import { isPast } from 'date-fns';
import { mongooseLoader } from '@entria/graphql-mongoose-loader';

import { DataLoaderKey, GraphQLContext, ObjectId } from '../../types';
import { isLoggedViewerCanSee, PLATFORM } from '../../security';

import { registerLoader } from '../loader/loaderRegister';

import { IUser } from '../user/UserModel';

import TokenModel, { IToken, TOKEN_SCOPES } from './TokenModel';

export default class Token {
  public registeredType = 'Token';

  id: string;
  _id: ObjectId;
  userId?: ObjectId;
  ip: string;
  scope: string;
  expiresIn: Date | null;
  isBlocked: boolean;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;

  constructor(data: IToken, context: GraphQLContext) {
    this.id = data.id || data._id;
    this._id = data._id;
    this.userId = data.userId;
    this.ip = data.ip;
    this.scope = data.scope;
    this.expiresIn = data.expiresIn;
    this.isBlocked = data.isBlocked;
    this.isActive = data.isActive;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
  }
}

export const getLoader = () => new DataLoader<DataLoaderKey, IToken>((ids) => mongooseLoader(TokenModel, ids));

registerLoader('TokenLoader', getLoader);

export const load = async (context: GraphQLContext, id: DataLoaderKey, isLogin?: boolean) => {
  if (!id) {
    return null;
  }

  try {
    const data = await context.dataloaders.TokenLoader.load(id);

    if (!data) {
      return null;
    }

    if (isLogin) {
      return new Token(data, context);
    }

    return isLoggedViewerCanSee(context, data) ? new Token(data, context) : null;
  } catch (err) {
    return null;
  }
};

interface LoadValidTokenProps {
  ctx: GraphQLContext;
  user: IUser;
  platform: PLATFORM;
  scope: TOKEN_SCOPES;
}

export const loadValidToken = async ({ ctx, user, platform, scope }: LoadValidTokenProps) => {
  const { koaContext } = ctx;
  const { ip } = koaContext.request;

  const token = await TokenModel.findOne({
    userId: user._id,
    ip,
    scope,
    platform,
    isBlocked: false,
    isActive: true,
  }).lean<IToken>();

  if (!token) {
    return null;
  }

  const hasExpired = token.expiresIn ? isPast(token.expiresIn) : false;

  if (hasExpired) {
    return null;
  }

  return new Token(token, ctx);
};
