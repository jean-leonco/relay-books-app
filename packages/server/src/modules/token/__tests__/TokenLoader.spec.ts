import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  sanitizeTestObject,
} from '@workspace/test-utils';

import { PLATFORM } from '../../../security';
import { createToken, createUser, getContext } from '../../../test/utils';

import * as TokenLoader from '../TokenLoader';
import { TOKEN_SCOPES } from '../TokenModel';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('TokenLoader', () => {
  it('should load null when looking for a invalid token', async () => {
    const user = await createUser();

    await createToken({ userId: user._id, isBlocked: true });

    const context = await getContext({ user });

    const validToken = await TokenLoader.loadValidToken({
      ctx: context,
      user,
      scope: TOKEN_SCOPES.SESSION,
      platform: PLATFORM.APP,
    });

    expect(validToken).toBeNull();
  });

  it('should only load me valid token', async () => {
    const user = await createUser();

    // invalid tokens
    for (let i = 0; i < 3; i++) {
      const notMeUser = await createUser();

      await createToken({ userId: notMeUser._id, isBlocked: i % 2 === 0, isActive: i % 2 !== 0 });
    }

    // valid tokens
    for (let i = 0; i < 3; i++) {
      const notMeUser = await createUser();

      await createToken({ userId: notMeUser._id });
    }

    await createToken({ userId: user._id });

    const context = await getContext({ user });

    const validToken = await TokenLoader.loadValidToken({
      ctx: context,
      user,
      scope: TOKEN_SCOPES.SESSION,
      platform: PLATFORM.APP,
    });

    expect(validToken?.userId).toEqual(user._id);
    expect(sanitizeTestObject(validToken)).toMatchSnapshot();
  });
});
