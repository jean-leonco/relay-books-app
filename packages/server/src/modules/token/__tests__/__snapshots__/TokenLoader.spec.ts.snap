// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenLoader should only load me valid token 1`] = `
Object {
  "_id": "ObjectId",
  "createdAt": "FROZEN-CREATEDAT",
  "expiresIn": "EMPTY",
  "id": "FROZEN-ID",
  "ip": "::ffff:127.0.0.1",
  "isActive": true,
  "isBlocked": false,
  "registeredType": "Token",
  "scope": "BOOKSAPP:SESSION",
  "updatedAt": "FROZEN-UPDATEDAT",
  "userId": "ObjectId",
}
`;
